web/client-api/src/main/java/io/deephaven/web/client/api/barrage/BarrageUtils.java

```diff
@@ -195,23 +196,30 @@ public DeltaUpdatesBuilder(BarrageUpdateMetadata barrageUpdate, boolean isViewpo
                 // if this isn't a viewport, then a second index isn't sent, because all rows are included
                 includedAdditions = deltaUpdates.getAdded();
             }
+            numAddRowsRemaining = includedAdditions.size();
             deltaUpdates.setIncludedAdditions(includedAdditions);
             deltaUpdates.setSerializedAdditions(new DeltaUpdates.ColumnAdditions[0]);
             deltaUpdates.setSerializedModifications(new DeltaUpdates.ColumnModifications[0]);
+
+            for (int columnIndex = 0; columnIndex < columnTypes.length; ++columnIndex) {
+                BarrageModColumnMetadata columnMetadata = barrageUpdate.modColumnNodes(columnIndex);
+                RangeSet modifiedRows = new CompressedRangeSetReader()
+                        .read(typedArrayToLittleEndianByteBuffer(columnMetadata.modifiedRowsArray()));
+                numModRowsRemaining = Math.max(numModRowsRemaining, modifiedRows.size());
+            }
         }
 
         /**
          * Appends a new record batch and payload. Returns true if this was the final record batch that was expected.
          */
         public boolean appendRecordBatch(RecordBatch recordBatch, ByteBuffer body) {
-            assert recordBatchesSeen < barrageUpdate.numAddBatches() + barrageUpdate.numModBatches();
-            if (barrageUpdate.numAddBatches() > recordBatchesSeen) {
+            if (numAddRowsRemaining > 0) {
                 handleAddBatch(recordBatch, body);
-            } else {
+            } else if (numModRowsRemaining > 0) {
                 handleModBatch(recordBatch, body);
             }
-            recordBatchesSeen++;
-            return recordBatchesSeen == barrageUpdate.numAddBatches() + barrageUpdate.numModBatches();
+            // return true when complete
+            return numAddRowsRemaining == 0 && numModRowsRemaining == 0;
```

Comment:
Is it even possible to have < 0? I wonder if we should prefer to throw IllegalStateException if we see that.